### PR TITLE
Create universal role and job for automated build/upload for OC images

### PR DESCRIPTION
**Pull request by fhubik**
Also, completely rewrite the old one, make this role
work with CI build and/or individual one (during OSPd deployment).

### CODE REVIEW

File: playbooks/build-oc-images.yml [hunk 1]
```diff
@@ -0,0 +1,51 @@
+---
```
Comment (yfried-redhat) [inline playbooks/build-oc-images.yml]: Please add a comment here (and in commit message) explaining that this playbook is not exposed via IR interface (ie `ir-installer`) but is rather intended for advance usage via `ansible-playbook` command

Comment (fhubik) [inline playbooks/build-oc-images.yml]: I believe this playbook doesn't fit to provision/install workflows and should be used only for building images, not actual deployments - therefore it shouldn't be included in CLI utils.

File: playbooks/installer/ospd/main.yml [hunk 2]
```diff
@@ -19,9 +19,17 @@
   hosts: undercloud
   gather_facts: no
   become: yes
-  become_user: "{{ installer.user.name }}"
+  become_method: sudo
   roles:
-      - {role: installer/ospd/images}
+      - role: installer/ospd/images
+        task: "{{ installer.images.task }}"
+        work_path: "{%- if installer.images.task == 'import' %}{{ '/home/{{ installer.user.name }}' }}{% endif %}"
```
Comment (yfried-redhat) [inline playbooks/installer/ospd/main.yml]: else?

File: playbooks/installer/ospd/main.yml [hunk 3]
```diff
@@ -19,9 +19,17 @@
...
+        ospd_version: "{{ installer.product.core.version }}"
```
Comment (yfried-redhat) [inline playbooks/installer/ospd/main.yml]: this is actually core (packstack version) so either you need `installer.product.version` or you need to change the name to `core_version`

File: roles/installer/ospd/images/defaults/main.yml [hunk 4]
```diff
@@ -1,10 +1,17 @@
...
 
-tar_images: "{{ installer.images.files | default(tar_files[installer.product.version]) }}"
+# Mandatory role parameters
+base_image_url: 'UNDEFINED'
+rhos_release_rpm_link: 'UNDEFINED'
+upload_to: 'none'
+storage_server: 'UNDEFINED'
+storage_path: 'UNDEFINED'
+storage_ssh_key: 'UNDEFINED'
+storage_user: 'UNDEFINED'
```
Comment (yfried-redhat) [inline roles/installer/ospd/images/defaults/main.yml]: No need to list mandatory args.
This defines defaults values as str("UNDFINED"/"none"). What we want is for ansible to throw errors in case of missing input

**Comment by fhubik** [comment]
retest openstack

**Comment by fhubik** [comment]
retest openstack

File: playbooks/installer/ospd/main.yml (hunk 2 above)
Comment (fhubik) [inline playbooks/installer/ospd/main.yml]: Else work_path will be taken from role defaults, as designed.

File: roles/installer/ospd/images/defaults/main.yml (hunk 4 above)
Comment (fhubik) [inline roles/installer/ospd/images/defaults/main.yml]: User may not have all mandatory params of certain role picked up from role's code, this way we can emphasize them a make role interface more clear - specifying mandatory params in defaults. Otherwise, without proper reading of code (and interpreting e.g. includes/conditionals) he may not notice at first look at "missing 'none'" whats wrong where. IMHO this really helps role interoperability.

File: playbooks/installer/ospd/main.yml (hunk 3 above)
Comment (fhubik) [inline playbooks/installer/ospd/main.yml]: Right, good point, I will fix in next patchset.

**Comment by fhubik** [comment]
Rebased.

**Comment by fhubik** [comment]
Poc's:

**Comment by krcmarik** [comment]
Hi,

I am experiencing that my jobs are repeatedly failing on images download from slow storage share (bos), one solution could be to raise a timeout for image download in roles/installer/ospd/images/tasks/import.yml but If this pull request rewrites the functionality and make it possible to have own built images on faster storage share then It would be great to get this merged asap. Thanks

File: playbooks/installer/ospd/main.yml (hunk 2 above)
Comment (yfried-redhat) [inline playbooks/installer/ospd/main.yml]: Not sure. IIUC, else workpath="". I think what you need here is ternary operator:

```yaml
work_path: {{ (installer.images.task == 'import') | ternary( '~' |expanduser, default(omit) ) }}
```

**Comment by krcmarik** [comment]
One more reason, It has been confirmed that the prebuilt images we are using are out of date and that's why we are not testing latest packages. The right way as It was said is to either use the prebuilt image and update them or maybe to build images agaisnt latest repo.

File: playbooks/installer/ospd/main.yml [hunk 5]
```diff
@@ -19,9 +19,17 @@
   hosts: undercloud
   gather_facts: no
   become: yes
-  become_user: "{{ installer.user.name }}"
+  become_method: sudo
```
Comment (tkammer) [inline playbooks/installer/ospd/main.yml]: if you have "become: yes" it defaults to "sudo" so no need for the "become_method: sudo" line

File: playbooks/build-oc-images.yml [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+            chmod 700 /root/.ssh/id_rsa
+
+- name: Build images as root and upload them to remote storage
+  hosts: controller
+  gather_facts: yes
+  become: yes
+  become_user: root
+  roles:
+    - role: installer/ospd/images
+      task: 'build'
```
Comment (tkammer) [inline playbooks/build-oc-images.yml]: instead of sending the params to the role, lets have everything in the role itself so a user can simply open the role and know what is being sent and from where.

**Comment by tkammer** [comment]
Overall looks great, but I would like to talk about this as I have some concerns looking forward + some new requirements that were asked from us.
ping me on IRC when you're ready.

File: playbooks/build-oc-images.yml (hunk 6 above)
Comment (fhubik) [inline playbooks/build-oc-images.yml]: If we specify our playbook-related variables an params directly in role, we would destroy role reusability completely and would bemaking mess for anyone who wants to work with the role independently. We would end-up in with variables like "private.something.something2" directly in role again, like in khaleesi.
Instead, I think is better to have default (and empty "mandatory") params speficied in role context (there are files "defaults/main.yml" and "vars/main.yml") and we are overriding them with our playbook-related/private variables on this level, this way we are not polluting role code with unrelated/private/external variables.

File: playbooks/installer/ospd/main.yml (hunk 5 above)
Comment (fhubik) [inline playbooks/installer/ospd/main.yml]: Ah, ok ;)

File: playbooks/installer/ospd/main.yml (hunk 2 above)
Comment (fhubik) [inline playbooks/installer/ospd/main.yml]: Ah, thanks, missed this, will fix in next commit.

File: playbooks/build-oc-images.yml (hunk 6 above)
Comment (mpavlase) [inline playbooks/build-oc-images.yml]: Why should be default values inside role pre-filled by private.somekey values? It's definitely not a good idea.
If I want to use something, that is designed to be reusable, it's my own business to pass proper values as I need to get desired behavior. It doesn't matter if it's me as 'user', or 'author' of this file. The author is kind of 'user' as well.
Role is like a 'function' in any ordinary language with ability to define arguments. If there would be set some default values, it should be chosen very carefully and not force user to redefine them to 'empty values to get basic functionality.

In short, I definitely agree with @fhubik and vote +1 for implementation in this patch.

**Comment by fhubik** [comment]
This PR is outdated and was moved to gerrithub for further work (https://review.gerrithub.io/#/c/278558/).
